sycl/source/detail/graph_impl.cpp

```diff
@@ -78,28 +78,162 @@ void duplicateNode(const std::shared_ptr<node_impl> Node,
+/// Propagation stops when an host task is encountered or when no predecessors
+/// remain.
+/// @param Node Node to assign to the partition
+/// @param PartitionNum Number to propagate
+void propagatePartitionUp(std::shared_ptr<node_impl> Node, int PartitionNum) {
+  if (((Node->MPartitionNum != -1) && (Node->MPartitionNum <= PartitionNum)) ||
+      (Node->MCGType == sycl::detail::CG::CGTYPE::CodeplayHostTask)) {
+    return;
+  }
+  Node->MPartitionNum = PartitionNum;
+  for (auto &Predecessor : Node->MPredecessors) {
+    propagatePartitionUp(Predecessor.lock(), PartitionNum);
+  }
+}
+
+/// Propagates the partition number `PartitionNum` to Successors
+/// Propagation stops when an host task is encountered or when no successors
+/// remain.
+/// @param Node Node to assign to the partition
+/// @param PartitionNum Number to propagate
+void propagatePartitionDown(
+    std::shared_ptr<node_impl> Node, int PartitionNum,
+    std::list<std::shared_ptr<node_impl>> &HostTaskList) {
+  if (Node->MCGType == sycl::detail::CG::CGTYPE::CodeplayHostTask) {
+    if (Node->MPartitionNum != -1) {
+      HostTaskList.push_front(Node);
+    }
+    return;
+  }
+  Node->MPartitionNum = PartitionNum;
+  for (auto &Successor : Node->MSuccessors) {
+    propagatePartitionDown(Successor.lock(), PartitionNum, HostTaskList);
+  }
+}
+
+/// Tests if the node is a root of its partition (i.e. no predecessors that
+/// belong to the same partition)
+/// @param Node node to test
+/// @return True is `Node` is a root of its partition
+bool isPartitionRoot(std::shared_ptr<node_impl> Node) {
+  for (auto Predecessor : Node->MPredecessors) {
+    if (Predecessor.lock()->MPartitionNum == Node->MPartitionNum) {
+      return false;
+    }
+  }
+  return true;
+}
+
 } // anonymous namespace
 
-void exec_graph_impl::schedule() {
+void partition::schedule() {
   if (MSchedule.empty()) {
-    for (auto &Node : MGraphImpl->MRoots) {
-      sortTopological(Node.lock(), MSchedule);
+    for (auto &Node : MRoots) {
+      sortTopological(Node.lock(), MSchedule, true);
     }
   }
 }
 
+void exec_graph_impl::makePartitions() {
+  int CurrentPartition = -1;
+  std::list<std::shared_ptr<node_impl>> HostTaskList;
+  // find HostTask
+  for (auto Node : MGraphImpl->MNodeStorage) {
+    if (Node->MCGType == sycl::detail::CG::CodeplayHostTask) {
+      HostTaskList.push_back(Node);
+    }
+  }
+
+  // annotate nodes
+  while (HostTaskList.size() > 0) {
+    auto Node = HostTaskList.front();
+    HostTaskList.pop_front();
+    CurrentPartition++;
+    for (auto &Predecessor : Node->MPredecessors) {
+      propagatePartitionUp(Predecessor.lock(), CurrentPartition);
+    }
+    CurrentPartition++;
+    Node->MPartitionNum = CurrentPartition;
+    CurrentPartition++;
+    auto tmpSize = HostTaskList.size();
+    for (auto &Successor : Node->MSuccessors) {
+      propagatePartitionDown(Successor.lock(), CurrentPartition, HostTaskList);
+    }
+    if (HostTaskList.size() > tmpSize) {
```

Comment:
This partition merging algorithm is complicated enough that it's worth documenting in code comments or design doc.